core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java

```diff
@@ -19,112 +19,205 @@
+    /** Bottom right corner coordinates of bounding box.*/
+    private GeoPoint bottomRight = null;
+    /** How the query should be run. */
+    private Type type;
 
-    private double[] box = {Double.NaN, Double.NaN, Double.NaN, Double.NaN};
-
-    private String type;
-
-    static final GeoBoundingBoxQueryBuilder PROTOTYPE = new GeoBoundingBoxQueryBuilder(null);
+    /**
+     * For serialization only.
+     * */
+    private GeoBoundingBoxQueryBuilder() {
+        this.fieldName = null;
+    }
 
-    public GeoBoundingBoxQueryBuilder(String name) {
-        this.name = name;
+    /**
+     * Create new bounding box query.
+     * @param fieldName name of index field containing geo coordinates to operate on.
+     * */
+    public GeoBoundingBoxQueryBuilder(String fieldName) {
+        this.fieldName = fieldName;
     }
 
     /**
      * Adds top left point.
-     *
      * @param lat The latitude
      * @param lon The longitude
      */
     public GeoBoundingBoxQueryBuilder topLeft(double lat, double lon) {
-        box[TOP] = lat;
-        box[LEFT] = lon;
+        topLeft = new GeoPoint(lat, lon);
         return this;
     }
 
+    /**
+     * Adds top left point.
+     * @param point point to add.
+     * */
     public GeoBoundingBoxQueryBuilder topLeft(GeoPoint point) {
-        return topLeft(point.lat(), point.lon());
+        this.topLeft = point;
+        return this;
     }
 
+    /**
+     * Adds top left point
+     * @param geohash GeoHash to derive the point from.
+     * */
     public GeoBoundingBoxQueryBuilder topLeft(String geohash) {
         return topLeft(GeoHashUtils.decode(geohash));
     }
 
+    /** Returns the top left corner of the bounding box. */
+    public GeoPoint topLeft() {
+        return topLeft;
+    }
+
     /**
      * Adds bottom right corner.
-     *
      * @param lat The latitude
      * @param lon The longitude
      */
     public GeoBoundingBoxQueryBuilder bottomRight(double lat, double lon) {
-        box[BOTTOM] = lat;
-        box[RIGHT] = lon;
+        this.bottomRight = new GeoPoint(lat, lon);
         return this;
     }
 
+    /**
+     * Adds bottom right corner.
+     * @param point the corner coordinates.
+     */
     public GeoBoundingBoxQueryBuilder bottomRight(GeoPoint point) {
-        return bottomRight(point.lat(), point.lon());
+        this.bottomRight = point;
+        return this;
     }
 
+    /**
+     * Adds bottom right corner.
+     * @param geohash the corner coordinates.
+     */
     public GeoBoundingBoxQueryBuilder bottomRight(String geohash) {
         return bottomRight(GeoHashUtils.decode(geohash));
     }
 
+    /** Returns the bottom right corner of the bounding box. */
+    public GeoPoint bottomRight() {
+        return bottomRight;
+    }
     /**
      * Adds bottom left corner.
      *
+     * @deprecated Use methods topLeft and bottomRight instead.
+     *
      * @param lat The latitude
      * @param lon The longitude
      */
+    @Deprecated
```

Comment:
Makes sense. Thanks for the pointer to OGC - I was actually searching for something like this to guide the refactoring.

Wrt. to updating the docs: When looking for a good spot on the bounding box query doc to add documentation about the lower left/ upper right I actually did find it mentioned in the paragraph on how to specify vertices (below the different formats we support). So it's there, I just didn't see it.